File: client/kotlin/nona-client/src/main/kotlin/com/nonaconfig/client/ValueParsing.kt
package com.nonaconfig.client

/**
 * Coercion from Nona's stored strings, checked against the shared mobile SDK
 * contract. Numbers use finite decimal syntax, without language-specific suffixes.
 */
internal object ValueParsing {
    private val integer = Regex("[+-]?[0-9]+")
    private val decimal = Regex("[+-]?(?:[0-9]+(?:\\.[0-9]*)?|\\.[0-9]+)(?:[eE][+-]?[0-9]+)?")

    fun parseBoolean(key: String, raw: String): ParseResult<Boolean> =
        when (raw.trim().lowercase()) {
            "true" -> ParseResult.Ok(true)
            "false" -> ParseResult.Ok(false)
            else -> mismatch(key, "boolean")
        }

    fun parseString(@Suppress("UNUSED_PARAMETER") key: String, raw: String): ParseResult<String> =
        ParseResult.Ok(raw)

    fun parseLong(key: String, raw: String): ParseResult<Long> =
        raw.trim().takeIf { integer.matches(it) }?.toLongOrNull()?.let { ParseResult.Ok(it) } ?: mismatch(key, "long")

    fun parseDouble(key: String, raw: String): ParseResult<Double> =
        raw.trim().takeIf { decimal.matches(it) }?.toDoubleOrNull()?.takeIf(Double::isFinite)?.let { ParseResult.Ok(it) }
            ?: mismatch(key, "double")

    private fun mismatch(key: String, type: String) = ParseResult.Err(
        NonaFailureReason.TYPE_MISMATCH,
        "Nona flag '$key' cannot be evaluated as a $type.",
    )

    sealed interface ParseResult<out T> {
        data class Ok<out T>(val value: T) : ParseResult<T>
        data class Err(val reason: NonaFailureReason, val message: String) : ParseResult<Nothing>
    }
}

File: client/javascript-openfeature-web-provider/src/value-parsing.ts
import type { NonaConfigValue } from "nona-client";

export type ParseFailureKind = "type-mismatch" | "parse-error";

export type ParseResult<T> =
  | { ok: true; value: T }
  | { ok: false; kind: ParseFailureKind; message: string };

export function parseBooleanValue(
  flagKey: string,
  config: NonaConfigValue,
): ParseResult<boolean> {
  const raw = config.value.trim().toLowerCase();
  if (raw === "true") {
    return { ok: true, value: true };
  }

  if (raw === "false") {
    return { ok: true, value: false };
  }

  return {
    ok: false,
    kind: "type-mismatch",
    message: `Nona flag '${flagKey}' cannot be evaluated as a boolean.`,
  };
}

export function parseStringValue(
  _flagKey: string,
  config: NonaConfigValue,
): ParseResult<string> {
  return { ok: true, value: config.value };
}

export function parseNumberValue(
  flagKey: string,
  config: NonaConfigValue,
): ParseResult<number> {
  const value = Number(config.value);
  if (config.value.trim() === "" || !Number.isFinite(value)) {
    return {
      ok: false,
      kind: "type-mismatch",
      message: `Nona flag '${flagKey}' cannot be evaluated as a number.`,
    };
  }

  return { ok: true, value };
}

export function parseObjectValue<T>(
  flagKey: string,
  config: NonaConfigValue,
): ParseResult<T> {
  try {
    return { ok: true, value: JSON.parse(config.value) as T };
  } catch {
    return {
      ok: false,
      kind: "parse-error",
      message: `Nona flag '${flagKey}' cannot be parsed as JSON.`,
    };
  }
}
